src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -5199,11 +5199,19 @@ private BoundExpression BindCollectionExpression(CollectionExpressionSyntax synt
 
             MessageID.IDS_FeatureCollectionExpressions.CheckFeatureAvailability(diagnostics, syntax, syntax.OpenBracketToken.GetLocation());
 
-            var builder = ArrayBuilder<BoundNode>.GetInstance(syntax.Elements.Count);
             foreach (var element in syntax.Elements)
             {
-                builder.Add(bindElement(element, diagnostics, this, nestingLevel));
+                if (element is KeyValuePairElementSyntax keyValuePairElement)
+                {
+                    MessageID.IDS_FeatureDictionaryExpressions.CheckFeatureAvailability(diagnostics, syntax, keyValuePairElement.ColonToken.GetLocation());
+                    return new BoundBadExpression(syntax, LookupResultKind.Empty, symbols: [], childBoundNodes: [], CreateErrorType());
```

Comment:
>return new BoundBadExpression [](http://example.com/codeflow?start=20&length=29)

We're not reporting any error here currently, although we are generating a BoundTree that cannot be emitted. Consider reporting an existing diagnostic, perhaps:
```csharp
diagnostics.Add(ErrorCode.ERR_ExpressionExpected, syntax.Location);
```